<a href="https://colab.research.google.com/github/hin1799/RS_Assignments/blob/main/RS_challenge2_Team5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Challenge 2

# Book recommendation system using collaborative filtering

**Submitted by:**

**Team 5**
**Team members**

1. Dhara Shah (202211008)
2. Hinal Desai (202211035)
3. Naiya Patel (202211075)
4. Jay Joshi (202211079)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install surprise #using surprise library for large matrix factorization

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 KB 13.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp39-cp39-linux_x86_64.whl size=3193678 sha256=69c1b0a99e06cb270abe57c9a9edc71e17e5838d9b920219667f7375b6c6d5ac
  Stored in directory: /root/.cache/pip/wheels/c6/3a/46/9b17b3512bdf283c6cb84f59929cdd5199d4e754d596d22784
Successfully built scikit-surprise


In [6]:
import os
import random
import operator
import requests
import numpy as np
import pandas as pd
from scipy import sparse
import sys
from surprise import Dataset, Reader
from surprise import KNNBasic, SVD
from surprise.model_selection import train_test_split
from surprise import accuracy
from surprise.dataset import DatasetAutoFolds
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
users_dir = "/content/drive/Shareddrives/LA-01/books/Users.csv"
books_dir = "/content/drive/Shareddrives/LA-01/books/Books.csv"
ratings_dir = "/content/drive/Shareddrives/LA-01/books/Ratings.csv"

In [4]:
users = pd.read_csv(users_dir)
books = pd.read_csv(books_dir)
ratings = pd.read_csv(ratings_dir)

<ipython-input-4-585a58d7bca1>:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(books_dir)


In [8]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [9]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [7]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [10]:
ratings['Book-Rating'].max()

10

In [11]:
ratings['Book-Rating'].min()

0

## Converting pandas dataframe to surprise dataset

In [12]:
reader = Reader(rating_scale=(0,10))
data = Dataset.load_from_df(ratings[['User-ID', 'ISBN', 'Book-Rating']], reader)

In [13]:
type(data)

surprise.dataset.DatasetAutoFolds

In [14]:
#Train test split
train_data, test_data = train_test_split(data, test_size=0.2)

## SVD

In [16]:
svd = SVD()
svd.fit(train_data)

In [17]:
#Making predictions
pred = svd.test(test_data)

In [18]:
pred

[Prediction(uid=100846, iid='0345447840', r_ui=0.0, est=1.089932764637708, details={'was_impossible': False}),
 Prediction(uid=135957, iid='0553583573', r_ui=10.0, est=4.337339903239933, details={'was_impossible': False}),
 Prediction(uid=242781, iid='0451184726', r_ui=0.0, est=3.697280804520864, details={'was_impossible': False}),
 Prediction(uid=166123, iid='0399140751', r_ui=8.0, est=3.309640451334454, details={'was_impossible': False}),
 Prediction(uid=153662, iid='0380818647', r_ui=9.0, est=1.1020989381905015, details={'was_impossible': False}),
 Prediction(uid=247447, iid='0446360716', r_ui=0.0, est=6.885510474702569, details={'was_impossible': False}),
 Prediction(uid=130474, iid='0872164683', r_ui=0.0, est=0.4104329948985401, details={'was_impossible': False}),
 Prediction(uid=79942, iid='0670865508', r_ui=0.0, est=1.6313099125367427, details={'was_impossible': False}),
 Prediction(uid=91342, iid='0843953314', r_ui=0.0, est=1.1843436080695613, details={'was_impossible': False})

pred contains the predictions as a list in the form of tuple (user, item, actual_rating, predicted_rating, details)

In [19]:
accuracy.rmse(pred)

RMSE: 3.5033


3.5033117362258546

In [21]:
accuracy.mae(pred)

MAE:  2.8192


2.8192363901708544

The RMSE value for rating prediction is 3.5033 and MAE is 2.8192.